nixos/modules/services/networking/ddclient.nix

```diff
@@ -184,26 +166,44 @@ with lib;
   ###### implementation
 
   config = mkIf config.services.ddclient.enable {
-    environment.etc."ddclient.conf" = {
-      enable = cfg.configFile == "/etc/ddclient.conf";
-      mode = "0600";
-      text = configText;
-    };
-
     systemd.services.ddclient = {
       description = "Dynamic DNS Client";
       wantedBy = [ "multi-user.target" ];
       after = [ "network.target" ];
-      restartTriggers = [ config.environment.etc."ddclient.conf".source ];
+      restartTriggers = optional (cfg.configFile != null) cfg.configFile;
 
-      serviceConfig = rec {
+      serviceConfig = {
         DynamicUser = true;
         RuntimeDirectory = StateDirectory;
         StateDirectory = builtins.baseNameOf dataDir;
         Type = "oneshot";
-        ExecStartPre = "!${lib.getBin pkgs.coreutils}/bin/install -m666 ${cfg.configFile} /run/${RuntimeDirectory}/ddclient.conf";
         ExecStart = "${lib.getBin pkgs.ddclient}/bin/ddclient -file /run/${RuntimeDirectory}/ddclient.conf";
       };
+      preStart = if cfg.configFile == null then ''
+        touch /run/${RuntimeDirectory}/ddclient.conf
+        chmod 0600 /run/${RuntimeDirectory}/ddclient.conf
+        cat >>/run/${RuntimeDirectory}/ddclient.conf <<EOF
+        # This file can be used as a template for configFile or is automatically generated by Nix options.
+        cache=${dataDir}/ddclient.cache
+        foreground=YES
+        use=${cfg.use}
+        login=${cfg.username}
+        ${optionalString (cfg.passwordFile != null) "password=$(head -n 1 ${cfg.passwordFile})"}
+        protocol=${cfg.protocol}
+        ${lib.optionalString (cfg.script != "") "script=${cfg.script}"}
+        ${lib.optionalString (cfg.server != "") "server=${cfg.server}"}
+        ${lib.optionalString (cfg.zone != "")   "zone=${cfg.zone}"}
+        ssl=${boolToStr cfg.ssl}
+        wildcard=YES
+        ipv6=${boolToStr cfg.ipv6}
+        quiet=${boolToStr cfg.quiet}
+        verbose=${boolToStr cfg.verbose}
+        ${cfg.extraConfig}
```

Comment:
Is everything going to escape nicely here? It might be worth writing the file out and then appending the password, like so:
```nix
let
  configFile = pkgs.writeText "ddclient.conf" ''
    # This file can be used as a template for configFile or is automatically generated by Nix options.
    cache=${dataDir}/ddclient.cache
    foreground=YES
    use=${cfg.use}
    login=${cfg.username}
    protocol=${cfg.protocol}
    ${lib.optionalString (cfg.script != "") "script=${cfg.script}"}
    ${lib.optionalString (cfg.server != "") "server=${cfg.server}"}
    ${lib.optionalString (cfg.zone != "")   "zone=${cfg.zone}"}
    ssl=${boolToStr cfg.ssl}
    wildcard=YES
    ipv6=${boolToStr cfg.ipv6}
    quiet=${boolToStr cfg.quiet}
    verbose=${boolToStr cfg.verbose}
    ${cfg.extraConfig}
    ${lib.concatStringsSep "," cfg.domains}
  '';
in
  preStart = ''
    install -m 600 ${configFile} /run/${RuntimeDirectory}/ddclient.conf
    ${optionalString (cfg.passwordFile != null) "password=$(head -n 1 ${cfg.passwordFile})"}
  '';
```

etc...

Or maybe the escaping won't be an issue :man_shrugging: What do you think?

Reply:
Good point, I updated the PR.